x-pack/plugins/snapshot_restore/common/constants.ts

```diff
@@ -19,3 +19,24 @@ export const PLUGIN = {
 };
 
 export const API_BASE_PATH = '/api/snapshot_restore/';
+
+export enum REPOSITORY_TYPES {
+  fs = 'fs',
+  url = 'url',
+  source = 'source',
+  s3 = 's3',
+  hdfs = 'hdfs',
+  azure = 'azure',
+  gcs = 'gcs',
+}
+
+export enum REPOSITORY_DOC_PATHS {
```

Comment:
I don't think this belongs in `common` since it's mean only to be consumed by client-side code. Can we move it into a constants file in the `public` directory?

After looking deeper, I think this applies to `REPOSITORY_TYPES` as well.

Reply:
I think `REPOSITORY_DOC_PATHS` should be moved to `public`, but there's a good chance that `REPOSITORY_TYPES` will be needed server-side at some point so I'm fine with keeping it in `common`

Reply:
My "_please remove all frictions_" brain tells me: why bother trying to separate constants file according to where they are consumed from. It adds a (small) layer of complexity that I don't think brings anything. When we import a constant we have to go to the following thought process:

- Is this constant used also on the server?
- Ummm I don't think so.... (So we go to the "public" folder, looking for the constant)
- Oh! I forgot, we updated last week and the server also requiresit  (and we have to go to the "common" folder check if it's indeed there)

Instead of:
- I need a constant... cool,  let's import it from `<app>/constants/index.ts`

No need to update anything if suddenly a constant changes context.

Reply:
I see what you're saying, Seb! From the point of view of someone writing code, it's easier to know all constants live in one place. In contrast, I think I tend to optimize for the DX of someone reading the code. Let me see if I can explain my perspective a bit, and why I think breaking things up makes the code simpler and easier to read.

Code written for the client has a different context than code written for the server, e.g. server-side code has access to Node APIs and client-side code has access to Web APIs, client-side code can use webpack aliases to import from `public/ui` and server-side code can't. Code which is written to be used both for the client and the server has to adhere to the restrictions of both server-side and client-side code, and can also by it's very nature be used in more places than that's written just for the server or the client. When I'm reading code, all of these "rules" factor into my mental model of how the code works, where it lives in the dependency tree, and where it will run.

Code organized into server-side or client-side buckets is easier for me to understand, because the domains are clearly defined and the rules for each are limited. `common` code is much more difficult for me because the domain is much broader, involves more rules, and is therefore more complex. It's sort of like putting variables into the global scope -- making changes to that code becomes more difficult because you're not sure what's using it.

Because of this, I'd strive to localize constants are closely as possible to where they're consumed. If constants are only used inside of a module to make it easier to read and maintain, then I'd keep them defined in that file. Once we discover a need to share them with another module, I'd extract them out into an area where they can be accessed more broadly, like a `constants` directory. By that same logic, once we discover a need to share them between both client and server, I'd move them into `common`. This way the location of the code communicates information about what may be using it.

Reply:
Thanks for detailing your view on this. Obviously, if there are rules about imports and so, it makes sense to not mix client with server code. I was referring to files simply exporting strings and values without any domain logic or references outside the scope of the file.

I guess we are in the "_move it until it feels right_" tweet that @jen-huang shared the other day 😊 As I get older (!) I tend to try to simplify the decision we're talking here. I can give you another example: when I was going through the rollup (and even the ccr) app, a few times I had to think: "in which folder is that component?" because of the way folder were structured. I had to remember who was its parent (in the DOM!), so I could go there and look for a subfolder to find the component. I understand "why" it's been put there, I used to do that a lot. Now I lean more towards removing all the need of thought process to find/import something and having a flatter structure when possible (as EUI does with its components).

You're going to ask me: "so when do we start to organize things into folders"? Well, that's the sweet spot that is difficult to get right, especially with multiple devs involved. In my personal project, I keep moving things around! 😄 so no wonder discussions like this one pop up from time to time in a team...